### ISSUE TITLE
Kernel error from dependent pattern match

### ISSUE THREAD

**Issue by david-christiansen** [leanprover/lean4#2588]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Check that your issue is not already [filed](https://github.com/leanprover/lean4/issues).
    * Reduce the issue to a minimal, self-contained, reproducible test case. Avoid dependencies to mathlib4 or std4.

### Description

While working on a complicated indexed representation of terms with binding, I ran into a pattern match that I thought should be accepted by Lean, but it was rejected with an error from the kernel.

On Zulip, @digama0 reduced it to this:
```
inductive Context : { _x : Bool // False } → Type where
  | empty (_ : False) : Context ⟨false, x⟩
  | extend : Context sc

def Context.lookup {sc : { _x : Bool // False }} (Γ : Context sc) : Unit :=
  match sc, Γ with
  | ⟨_, _⟩, Context.extend => ()
```
The error is:
```
(kernel) function expected
  _x_6 (_ : HEq Γ✝ Γ✝)
```

### Context

[Discussion on Lean Zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Is.20this.20a.20bug.20in.20the.20pattern.20compiler.3F)

### Steps to Reproduce

1.Enter the code in a buffer
2. Observe the error message

**Expected behavior:**

The code is accepted

**Actual behavior:** 

A red underline on the `match sc ...` line, with the given error message

### Versions

```
$ lean --version
Lean (version 4.2.0-rc1, commit a62d2fd49796, Release)
$ uname -a
Linux localhost.localdomain 6.4.15-200.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Sep  7 00:25:01 UTC 2023 x86_64 GNU/Linux
```

### Additional Information

None

### Impact

Add :+1: to [issues you consider important](https://github.com/leanprover/lean4/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc). If others are impacted by this issue, please ask them to add :+1: to it.

**Comment by enricozb** [comment]
I ran into a similar issue, but I'm unsure if it is the same bug:
```
import Mathlib.Data.Finset.Basic

inductive Affine : (vs : Finset ℕ) → Type
| var (x : ℕ) : Affine {x}
| abs (x : ℕ) (e : Affine vs) : Affine (vs \ {x})
| app (e₁ : Affine vs₁) (e₂ : Affine vs₂) (h : vs₁ ∩ vs₂ = ∅) : Affine (vs₁ ∪ vs₂)

/-
  (kernel) function expected
  _x_24 (_ : HEq e₁✝ e₁✝)
-/
def small_step (e : Affine vs) : (vs' : Finset ℕ) × Affine vs' :=
  match e with
  | .app (.abs x e₁) e₂ h =>
    if x ∈ vs then
      ⟨_, e₂⟩
    else
      ⟨_, e₁⟩
  | e => ⟨_, e⟩
```

**Comment by zwarich** [comment]
The original example appears to be fixed on `master`.

**Comment by enricozb** [comment]
@zwarich, indeed. Both the original and my example seem to be fixed on `master`.

**Comment by nomeata** [comment]
Likely a duplicate of #1774, which has more examples, some fixed and some not.
